### PR TITLE
8290047: (fs) FileSystem.getPathMatcher does not check for ":" at last index

### PR DESCRIPTION
**Pull request by bplb**
For a `String` “s”, `s.indexOf(int)` can never return a value `>= s.length()` so change the check
```
        int pos = syntaxAndInput.indexOf(':');
        if (pos <= 0 || pos == syntaxAndInput.length())
```
to
```
        if (pos <= 0)
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8290047](https://bugs.openjdk.org/browse/JDK-8290047): (fs) FileSystem.getPathMatcher does not check for ":" at last index
 * [JDK-8291730](https://bugs.openjdk.org/browse/JDK-8291730): (fs) FileSystem.getPathMatcher does not check for ":" at last index (**CSR**)

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [291ef674](https://git.openjdk.org/jdk/pull/9595/files/291ef674e6d1975f94bef9e498f89fa34a1ee0eb)
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [b9de2604](https://git.openjdk.org/jdk/pull/9595/files/b9de2604b74c1082c6b0fa10435405120b971e93)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [b9de2604](https://git.openjdk.org/jdk/pull/9595/files/b9de2604b74c1082c6b0fa10435405120b971e93)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [b9de2604](https://git.openjdk.org/jdk/pull/9595/files/b9de2604b74c1082c6b0fa10435405120b971e93)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9595/head:pull/9595` \
`$ git checkout pull/9595`

Update a local copy of the PR: \
`$ git checkout pull/9595` \
`$ git pull https://git.openjdk.org/jdk pull/9595/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9595`

View PR using the GUI difftool: \
`$ git pr show -t 9595`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9595.diff">https://git.openjdk.org/jdk/pull/9595.diff</a>

</details>

### CODE REVIEW

**Review by naotoj — APPROVED** [review]
Looks good, Brian.

File: src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java [hunk 1]
```diff
@@ -175,7 +175,7 @@ public final String getSeparator() {
     @Override
     public PathMatcher getPathMatcher(String syntaxAndInput) {
         int pos = syntaxAndInput.indexOf(':');
-        if (pos <= 0 || pos == syntaxAndInput.length()) {
```
Comment (RogerRiggs) [inline src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java:178]: Is this really a different bug? Should it be checking for `length() - 1` to throw IAE.
That would then throw if there was no string after the syntax selector;  for example "glob:"

Comment (bplb) [reply src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java:178]: You have a point there. It does not want the ":" to be at the first position so it should not want it at the last position as the "pattern" part of the string would be empty. That fits with the specification
```
IllegalArgumentException.html - If the parameter does not take the form: syntax:pattern
```

Comment (bplb) [reply src/java.base/share/classes/jdk/internal/jrtfs/JrtFileSystem.java:178]: Corrected in 5765fa1c6883575da57499c7b70f6626443d869d.

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java [hunk 2]
```diff
@@ -741,10 +741,12 @@ <V extends Number,A> Future<V> implWrite(boolean isGatheringWrite,
                 return result;
             }
         } catch (Throwable x) {
+            System.out.println("Caught " + x);
```
Comment (naotoj) [inline src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java:744]: Leftover debug output?

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java [hunk 3]
```diff
@@ -741,10 +741,12 @@ <V extends Number,A> Future<V> implWrite(boolean isGatheringWrite,
...
             if (x instanceof ClosedChannelException)
                 x = new AsynchronousCloseException();
             exc = x;
         } finally {
+            System.out.println("finally");
```
Comment (naotoj) [inline src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java:749]: As above

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java (hunk 2 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java:744]: Yes, thanks for catching.

**Review by RogerRiggs — CHANGES_REQUESTED** [review]
Please update the issue title and description and PR title and description to match the new understanding.  

The IAE message in JrtFileSystem wording is a bit like it is leftover debugging output.
The messages for all 4 IAE could be something as simple as "missing pattern".

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java (hunk 3 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java:749]: This file was a leftover from a different issue I was looking at and was accidentally included here.

**Comment by bplb** [comment]
> The IAE message in JrtFileSystem wording is a bit like it is leftover debugging output. The messages for all 4 IAE could be something as simple as "missing pattern".

Not sure a message is needed. At the other extreme one could say "missing syntax" for ":pattern" and "missing pattern" for "syntax:". Or it could be "Parameter does not take the form: syntax:pattern".

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java (hunk 2 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java]: Fixed in ff404d4b52c5ecf4f00743be1d4d4a266483460c.

File: src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java (hunk 3 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/ch/UnixAsynchronousSocketChannelImpl.java]: Fixed in ff404d4b52c5ecf4f00743be1d4d4a266483460c.

**Comment by bplb** [comment]
Is there anything else to be done in this PR? Thanks.

**Comment by AlanBateman** [comment]
> Is there anything else to be done in this PR? Thanks.

It looks like this issue has changed from the original bug report and now rejects the empty pattern with IAE. This is a behavior change (and probably the right behavior) but I think will require a small update to the javadoc to make it clear that both an empty syntax and empty pattern are illegal.

**Comment by bplb** [comment]
> [...] I think will require a small update to the javadoc to make it clear that both an empty syntax and empty pattern are illegal.

I read the existing specification
```
     * @throws  IllegalArgumentException
     *          If the parameter does not take the form: {@code syntax:pattern}
```
as covering both cases, but I suppose one could explicitly state that "neither syntax nor pattern may be empty."

**Comment by bplb** [comment]
> Alternatively, change "where ':' stands for itself" to "where syntax and pattern components are non-empty strings and ':' stands for itself" so that it's up front when the input string is described.

I think that reads better; will change accordingly.

**Comment by RogerRiggs** [comment]
> 

The existing code downstream of the check handles an empty glob or regex pattern as matching the empty string.
If I read it correctly, it will now throw an exception instead of not matching.  It might be safer to not change that behavior.

**Comment by bplb** [comment]
> The existing code downstream of the check handles an empty glob or regex pattern as matching the empty string. If I read it correctly, it will now throw an exception instead of not matching. It might be safer to not change that behavior.

Indeed for this code
```
        FileSystem fs = ...;
        PathMatcher pm = fs.getPathMatcher("glob:");
        System.out.println(pm.matches(Path.of("")));
```
without the proposed change `true` is printed but with the change an `IllegalArgumentException` is thrown so it would be safer to change
```
        if (pos <= 0 || pos == syntaxAndInput.length())
```
to
```
        if (pos <= 0)
```
as originally suggested.

**Comment by AlanBateman** [comment]
> The existing code downstream of the check handles an empty glob or regex pattern as matching the empty string.
> If I read it correctly, it will now throw an exception instead of not matching. It might be safer to not change that behavior.

I'm 99% sure the original intention was for the empty pattern to be rejected with IAE. It was an implementation but that it checked if the index is length rather than length-1. That was JDK 7 and changing is now without be behavior change. I can't imagine the impact would be significant but I agree it's not worth it and maybe we just drop the check as per the JBS issue and update the spec to document the long standing behavior.

**Comment by bplb** [comment]
> [...] I agree it's not worth it and maybe we just drop the check as per the JBS issue and update the spec to document the long standing behavior.

Agreed.

File: src/java.base/share/classes/java/nio/file/FileSystem.java [hunk 4]
```diff
@@ -304,7 +304,7 @@ protected FileSystem() {
      * <blockquote><pre>
      * <i>syntax</i><b>:</b><i>pattern</i>
      * </pre></blockquote>
-     * where {@code ':'} stands for itself.
+     * where <i>syntax</i> must not be empty and {@code ':'} stands for itself.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/nio/file/FileSystem.java:307]: We could flesh this out a bit more with something like:

"where <i>syntax</i> is the non-empty name of the syntax,  <i>pattern</i> is a possibly-empty pattern string, and {@code ':'} stands for itself."

Comment (bplb) [reply src/java.base/share/classes/java/nio/file/FileSystem.java:307]: That is better.

Comment (bplb) [reply src/java.base/share/classes/java/nio/file/FileSystem.java]: So changed in b9de2604b74c1082c6b0fa10435405120b971e93.

**Review by AlanBateman — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]

**Review by LanceAndersen — APPROVED** [review]

**Pull request closed without merging**